### PR TITLE
Anagh davide calle create parsed dataset

### PR DESCRIPTION
**Pull request by pyg4thr1x**

### CODE REVIEW

**Comment by hcw1026** [comment]
Comments:
**_tid_filter**
- tf.math.reduce_any (probably doesn't matter)

**_tag_filter**
- tf.math.equal (same as above) and other similar tf.*
- True in tags_mask implies True in tags_bool, so there is no need to compute tags_bool?

-Big-Problem:
- indices count from zero, but tags count from one

**_tag_filter_hotenc_mask**
- tf.boolean_mask is deprecated

- Big-Problem:
indices count from zero, but tags count from one

**generate_dataset**
- any is a special word, choose another name?
- add an option for no slicing?
- how can I choose the division between train_test_val? and how can I choose subset of tfrecords for testing train.py (e.g. I just want to use two files with one epoch)



**Incompatible with model_keras.py and train.py**
- 'waveform' does not appear in audio_processing, and I use 'log-mel-spectrogram' and 'waveform' in model_keras as well, so which one to use?
- I can't use sparse tensor in train.py directly
- boolean mask shape is different to the output of model_keras, (either change this or change model_keras)
- dimension of tags are preferrabilly (-1, 155)


DONE

**Comment by pyg4thr1x** [comment]
All done. Also fixed vectorization of _window function. generate_dataset() should give the right output. I just want to add some batch normalization, and then for me we're ready to go

**Comment by pyg4thr1x** [comment]
Done

**Review by callesonne — COMMENTED** [review]
Review
----------
Overall extremely well made code, Im very happy with it.

audio_processing:
- I'm not sure why you changed the default database to the val!=0 one, as I have already run the code using the one ignoring val.. So if someone were to replicate the code we have now they would get different results from ours.

projectName_input:

_parse_features:
- Im not sure if I like the fact that audio_feature_description is a global variable. This is probably nitpicking but I really like the way you solved the reshaping problem, i.e. define TENSOR_SHAPE in the main function and then nest the reshape function within a lambda function to make it compatible with dataset.map AND can have multiple input arguments.

Why are we doing batch normalization in the generate dataset function? Because from what I know you are not supposed to do batch normalization on the test data so this makes generate_dataset only usable on train data?

**Comment by hcw1026** [comment]
1. tf.nn.sigmoid ..... takes float32 instead of int
2. reshaping waveform, what (-1,), should it be (1,-1)? or no reshaping at all?

**Comment by hcw1026** [comment]
I am happy with the script once the changes are made.
Do you want to move the generate_datasets from train.py to this script or to model.py? or leave that in train.py? If you want to move to this script, please do.

**Comment by pyg4thr1x** [comment]
Overview of my changes:
- Fix 'genrate' typo, self explanatory...
- Fix old audio_processing bug (error while print), not related to my module - I found an error while trying to run audio_processing locally, now fixed
- Add batch tuplification, what I mentioned earlier, now generate_dataset outputs a tuple ready for model.fit
- Change variable names, ... (forgive me)
- Fix default database, answer to Calle's first comment
- Features description now local variable, answer to Calle's second comment

**Comment by pyg4thr1x** [comment]
> 1. tf.nn.sigmoid ..... takes float32 instead of int

Don't get it... What are you referring to?

**Comment by pyg4thr1x** [comment]
> 2\. reshaping waveform, what (-1,), should it be (1,-1)? or no reshaping at all?

(-1, ) does no reshaping at all

**Comment by pyg4thr1x** [comment]
> Do you want to move the generate_datasets from train.py to this script or to model.py? or leave that in train.py? If you want to move to this script, please do.

No, I wouldn't move your generate_datasets to this script. Maybe to model.py, just to avoid repeating code within the custom training module and the standard training model... Still, it's beyond the scope of this PR I think

**Comment by pyg4thr1x** [comment]
> I am happy with the script once the changes are made.

I didn't fully get your two comments... I've done what I still felt needed to be done, feel free to tweak the code as you wish (as long as inline comments start with lowercase letter and don't end with fullstop 😉😝)

**Comment by callesonne** [comment]
I dont see a change on the audio_processing bug commit, Just improvement of documentation + variable renaming. Am I missing something?

**Comment by callesonne** [comment]
Why were we doing batch normalisation in this script? The test data shouldnt be normalised right?

**Comment by hcw1026** [comment]
> > 1. tf.nn.sigmoid ..... takes float32 instead of int
> 
> Don't get it... What are you referring to?

e.g. tf.nn.softmax cannot accept [1, 0], it can only accept [1., 0.]

**Comment by pyg4thr1x** [comment]
> I dont see a change on the audio_processing bug commit, Just improvement of documentation + variable renaming. Am I missing something?

There was a `print(name) + '.tfrecord'` - that is, string concatenation outside the print function (just a typo, but the script clearly returned an error when dealing with the last file)

**Comment by pyg4thr1x** [comment]
> e.g. tf.nn.softmax cannot accept [1, 0], it can only accept [1., 0.]

I see... I'll fix that

**Comment by pyg4thr1x** [comment]
> Why were we doing batch normalisation in this script? The test data shouldnt be normalised right?

Not sure... Just add an option not to normalize?

**Review by callesonne — APPROVED** [review]
I approve of this now, again good work!
